app/modules/github.js

```diff
@@ -8,37 +8,46 @@ define(['knockout', 'durandal/system', 'plugins/http', './node'], function(ko, s
       'Accept' : 'application/vnd.github.v3+json'
       };
 
-    function Github() {}
-
-    Github.prototype.fetchRepository = function(owner, repo) {
+    function Github() {
+        this.repositories = {};
+    }
 
+    Github.prototype.fetchRepository = function(owner, repo, sha1) {
+      var self = this;
       // todo: the desired sha should be provided by
       // the owner requesting the feedback
 
-        var sha1 = 'master';
+        if (sha1 && sha1 !== '' && self.repositories[sha1]) {
+            return system.defer(function(cb) { cb.resolve(self.repositories[sha1]); });
+        }
+
+        sha1 = sha1 || 'master';
         var url = base_url + 'repos/' + owner + '/' + repo + '/git/trees/' + sha1 + '?recursive=1';
 
         return http.get(url, null, base_headers).then(function(response, status, xhr) {
             if (response.truncated) { /* signal error condition */ }
             var root = Node.unflatten(response.tree);
             root.sha = response.sha;
+            root.shortSha = response.sha.substring(0,7);
+            self.repositories[root.sha] = root;
+            self.repositories[root.shortSha] = root;
             return root;
         });
     };
 
-    Github.prototype.fetchRawFile = function(owner, repo, sha1) {
+    Github.prototype.fetchRawFile = function(owner, repo, commitSha1, filePath) {
+        return this.fetchRepository(owner, repo, commitSha1).then(function(repository) {
+            var file = repository.lookupFileByPath(filePath);
+            var url = base_url + 'repos/' + owner + '/' + repo + '/git/blobs/' + file.blobSha();
 
-      var url = base_url + 'repos/' + owner + '/' + repo + '/git/blobs/' + sha1;
-
-      var headers = {
-        'Accept' : 'application/vnd.github.v3+json'
-      };
-
-      return http.get(url, null, base_headers).then(function(response, status, xhr) {
-        var decoded = atob(response.content);
-        return decoded;
-      });
+            return http.get(url, null, base_headers).then(function(response, status, xhr) {
+                var decoded = atob(response.content);
+                return decoded;
+            });
+        });
     };
 
+    Github.instance = new Github();
```

Comment:
The intent is to have a singleton instance across the app, right? We can simplify this code then.
I'll address that in a later PR.